### PR TITLE
Cleaning up mode mqtt delegate

### PR DESCRIPTION
**Pull request by shigmas**
Updates for integrating as a Mode client. Two major changes:
 - adding a republish
 - exposing some properties

Details:
For using as a replacement of the SDK:
 - made the construction of the ModeMqttDelegate more uniform
 - use the simplified functional options for the ModeMqttDelegate (rather than the crazy one) 
 - delegate now creates downstream channels, since we close them
 - exposed the subscription handler function type.
 - some error checking
 - handle if there is no server when we create a connection.
 - removed requestTimeout since we weren't using it because we rely on context.Context
 -  created a Republish method, which takes a packetID. The original publish is mostly wrapped by Publish and Republish. If a packet ID is passed in (Republish), we reuse that.
 - updated the dummy MQTT Server to send back some KV updates. So, removed the kvCmd and replaced it with kvSync, kvSet, and KvDelete.

### CODE REVIEW

File: v3/mqtt.go [hunk 1]
```diff
@@ -437,7 +440,16 @@ func (client *MqttClient) PingAndWait(ctx context.Context) error {
...
+	return client.publishWithID(ctx, qos, topic, data, 0)
+}
 
+func (client *MqttClient) Republish(ctx context.Context, qos QOSLevel,
+	topic string, data []byte, packetID uint16) (uint16, error) {
+	return client.publishWithID(ctx, qos, topic, data, packetID)
+}
+
+func (client *MqttClient) publishWithID(ctx context.Context, qos QOSLevel,
+	topic string, data []byte, packetID uint16) (uint16, error) {
```
Comment (shigmas) [inline v3/mqtt.go:452]: 2 wrappers over publishWithID, and line 676 is where we check it.

File: v3/mqtt_dummy.go [hunk 2]
```diff
@@ -212,29 +237,38 @@ func sendPublish(client *dummyClient, pubCmd DummyCmd) {
 				Payload: globalConf.SubData,
 				QOS:     packet.QOSAtMostOnce,
 			}
-			pubPkt.PacketID = 1
+			pubPkt.PacketID = uint16(atomic.AddUint32(&packetID, 1))
 			pkts = append(pkts, pubPkt)
 		}
-	case PublishKvCmd:
+	case PublishKvSync:
```
Comment (shigmas) [inline v3/mqtt_dummy.go:243]: Replaced the single Cmd (which was sync), to Sync, Set, andDelete

Comment (shigmas) [inline v3/mqtt_dummy.go:243]: Instead of one command to publish a KV Sync, we now have KVSync, KVSet, and KVDelete.

File: v3/mqtt_test.go [hunk 3]
```diff
@@ -283,31 +283,54 @@ func TestMqttClientPublish(t *testing.T) {
 	DummyMQTTD(ctx, &wg, nil)
 	fmt.Println("TestMqttClientPublish")
 
-	client := testConnection(ctx, t, delegate, false)
+	t.Run("Test Publish With ACK", func(t *testing.T) {
```
Comment (shigmas) [inline v3/mqtt_test.go:286]: If you hide whitespace changes, there's only one new section (line 308...)

**Review by takjn — APPROVED** [review]
LGTM

File: v3/mqtt_dummy.go [hunk 4]
```diff
@@ -38,9 +41,16 @@ var (
...
+	DefaultItems = []*KeyValue{
+		&KeyValue{Key: "key1", Value: "value1", ModificationTime: time.Now()},
+		&KeyValue{Key: "key2", Value: "value2", ModificationTime: time.Now()},
+	}
+
 	currentDevice = "" // XXX - fix the logic that stores this globally
 
 	globalConf DummyConfig
+	packetID   uint32 = 1
+	kvRevision uint32 = 1
```
Comment (takeshinoda) [inline v3/mqtt_dummy.go:53]: Ah global variable, So it needs atomic operation.

File: v3/mqtt_dummy.go [hunk 5]
```diff
@@ -38,9 +41,16 @@ var (
...
 
+	DefaultItems = []*KeyValue{
+		&KeyValue{Key: "key1", Value: "value1", ModificationTime: time.Now()},
+		&KeyValue{Key: "key2", Value: "value2", ModificationTime: time.Now()},
+	}
+
 	currentDevice = "" // XXX - fix the logic that stores this globally
 
 	globalConf DummyConfig
+	packetID   uint32 = 1
```
Comment (takeshinoda) [inline v3/mqtt_dummy.go:52]: Why it this variable `uint32`? Packet id specification is `uint16`, I think.

**Review by takeshinoda — APPROVED** [review]
LGTM!!

File: v3/mqtt_dummy.go (hunk 5 above)
Comment (shigmas) [reply v3/mqtt_dummy.go:52]: You're right, but atomic only has add for 32 and 64.

File: v3/mqtt_dummy.go (hunk 4 above)
Comment (shigmas) [reply v3/mqtt_dummy.go:53]: Yes, not so good. This was a simple test server, but it kept getting bigger. Maybe next change, I will refactor it.

**Pull request merged**